Speedup movegen with VBMI2

87flowers opened the pull request:
Passed STC
LLR: 2.96 (-2.94,2.94) <0.00,2.00>
Total: 166720 W: 43191 L: 42701 D: 80828
Ptnml(0-2): 348, 18567, 45069, 18999, 377 
https://tests.stockfishchess.org/tests/view/686ae98dfe0f2fe354c0c867

-----

Refactor movegen to emit to a vector with 16-bit elements, which enables a speedup with AVX512-VBMI2 when writing moves to the move list.

Very crude timing measurements of perft via timing `./stockfish "go perft 7"` demonstrates approximately 17% perft speedup:
```
Summary
  ./Stockfish-dev/src/stockfish 'go perft 7' ran
    1.17 ± 0.04 times faster than ./Stockfish-base/src/stockfish 'go perft 7'
```

Estimated overall nps increase of 0.4% via `speedtest`: `33605229` -> `33749825` (many thanks @JonathanHallstrom).

------

Explanation of change:

During movegen, you often "splat" a bitboard's worth of moves into the movelist. In stockfish, this is from a `from` square to a `to_bb` bitboard, like so:

```
    while (to_bb)
        *moveList++ = Move(from, pop_lsb(to_bb));
```

This patch speeds that up by replacing the splat operation with two [`vpcompressw` instructions](https://www.officedaytime.com/simd512e/simdimg/compress.php?f=vpcompressw).

To use pawn movegen as an example, you generate preconstructed list of moves in two 512-bit vectors (a total of 64 elements), like: `a1a2`, `b1b2`, `c1c2`, ..., `a2a3`, `b2b3`, `c2c3`, .... `h6h7`, ....
We then use a bitboard to select which of these moves are actually valid, and write the valid moves into the move list.

For regular moves, `SPLAT_TABLE` is a vector that looks like `a1a1`, `a1a2`, `a1a3`, ..., `a1h8`. Notice that the from square for this vector of moves is 0. This allows us to later `or` in the `from` square by constructing a from vector with `_mm512_set1_epi16(from << 6)`, which we `_mm512_or_si512` into `SPLAT_TABLE` before splatting out the valid moves with the `to_bb` bitboard.

This is faster, because doing the splat as single instruction avoids the highly unpredictable bitboard loop, which CPUs do not like.

87flowers commented on src/movegen.cpp:
```diff
@@ -24,12 +24,93 @@
+}
+
+inline Move* splat_moves(Move* moveList, Square from, Bitboard to_bb) {
+    __m512i fromVec = _mm512_set1_epi16(from << 6);
```
Maybe instead:
```suggestion
    __m512i fromVec = _mm512_set1_epi16(Move(from, SQUARE_ZERO).raw());
```

Disservin commented:
a bit of a write explaining how this works in the pr description wouldn't hurt ^^

Sopel97 commented:
What CPU was it benchmarked on? vpcompressw is 256 uops on zen 4 so it could be disastrous

87flowers commented:
> What CPU was it benchmarked on? vpcompressw is 256 uops on zen 4 so it could be disastrous

The perft test was done on a zen 4 machine. The speedtest was done on a zen 5 machine.
The 256 ops variant was explicitly avoided in this patch! A workaround was done to avoid that specific instruction.

Sopel97 commented:
interesting, looks like only the variant that loads from memory is like that, I wonder why

mstembera commented:
> > Also not sure we need x86-64-vbmi2-256 and x86-32-vbmi2-256 at all.
> 
> I was just following existing practice of having 256 bit wide registers only by default.

x86-64-vnni256 was originally only added because it was slightly faster on the first generation cpus supporting vnni because they down clocked significantly when using 512 wide registers (https://github.com/official-stockfish/Stockfish/pull/6153).  Since then we had many avx512 optimizations (https://github.com/official-stockfish/Stockfish/pull/6153 https://github.com/official-stockfish/Stockfish/pull/6139 among others) that I believe this is no longer the case and x86-64-vnni256 could/should be deprecated.  x86-64-avxvnni is probably still useful as a 256 wide vnni arch though.

mstembera commented:
So I ran a standard bench  (not perft) and measure no speed difference.  I hope more people can bench so we know better what if any difference this makes in normal playing scenarios we actually care about.

mstembera commented on src/movegen.cpp:
```diff
@@ -24,12 +24,94 @@
+static_assert(sizeof(SPLAT_TABLE) == 128);
+
+inline Move* write_moves(Move* moveList, uint32_t mask, __m512i vector) {
+    _mm256_storeu_si256((__m256i*) moveList,
```
You can probably use _mm512_mask_compressstoreu_epi16() here.

mstembera replied:
Ah ok.  We had similar issues with the pext instruction on Zen2 and prior.

mstembera commented on src/movegen.cpp:
```diff
@@ -24,12 +24,94 @@
+inline Move* write_moves(Move* moveList, uint32_t mask, __m512i vector) {
+    _mm256_storeu_si256((__m256i*) moveList,
+                        _mm512_castsi512_si256(_mm512_maskz_compress_epi16(mask, vector)));
+    return moveList + __builtin_popcount(mask);
```
Could you change this to ```return moveList + popcount(mask);``` so it also compiles under MSVC.

MinetaS commented:
Tested on AMD Ryzen 9 9950X3D

master (ce73441f2013e0b8fd3eb7a0c9fd391d52adde70)
```
Version                    : Stockfish dev-20250702-ce73441f
Compiled by                : clang++ 19.1.7 on Linux
Compilation architecture   : x86-64-vnni512
Compilation settings       : 64bit VNNI AVX512 BMI2 AVX2 SSE41 SSSE3 SSE2 POPCNT
Compiler __VERSION__ macro : Ubuntu Clang 19.1.7 (3ubuntu1)
Large pages                : yes
User invocation            : speedtest 
Filled invocation          : speedtest 32 4096 150
Available processors       : 0-31
Thread count               : 32
Thread binding             : none
TT size [MiB]              : 4096
Hash max, avg [per mille]  : 
    single search          : 59, 29
    single game            : 757, 550
Total nodes searched       : 7646846920
Total search time [s]      : 153.515
Nodes/second               : 49811724
```

87flowers/move-splat-2 (19fc233ea2833b56873cf4d6df5bdf1eb104592b)
```
Version                    : Stockfish dev-20250710-19fc233e
Compiled by                : clang++ 19.1.7 on Linux
Compilation architecture   : x86-64-vbmi2-512
Compilation settings       : 64bit VBMI2 VNNI AVX512 BMI2 AVX2 SSE41 SSSE3 SSE2 POPCNT
Compiler __VERSION__ macro : Ubuntu Clang 19.1.7 (3ubuntu1)
Large pages                : yes
User invocation            : speedtest 
Filled invocation          : speedtest 32 4096 150
Available processors       : 0-31
Thread count               : 32
Thread binding             : none
TT size [MiB]              : 4096
Hash max, avg [per mille]  : 
    single search          : 60, 29
    single game            : 761, 548
Total nodes searched       : 7673171842
Total search time [s]      : 153.515
Nodes/second               : 49983205
```

Note that I got >50Mnps from master at the first iteration (as a warm up), so I'd say the speedup is either marginal or does not exist.

Also `x86-64-vbmi2-*` presets are missing in `make help` text.

mstembera commented:
> Edit. It seems the test will have passed with +0.7-0.9 Elo, which may suggest the speedup is +0.4%, congruous to my result.

However, the test with the current improved version actually failed https://tests.stockfishchess.org/tests/view/686c15301451bd6fb83073d3

87flowers commented:
updated PR to drop the "improved version" commit that failed

87flowers commented:
```
LLR: 2.96 (-2.94,2.94) <0.00,2.00>
Total: 166720 W: 43191 L: 42701 D: 80828
Ptnml(0-2): 348, 18567, 45069, 18999, 377 
```

cj5716 commented:
A side note, we should probably use reinterpret_cast instead of C casts to follow the convention in the codebase

87flowers commented:
Addressed cj's comment; and also addressed mstembera's comments regarding renaming the architecture.

Renamed to avx512icl because i feel that's the most reflective name and [it is a good baseline for consumer avx-512 feature set support](https://en.wikipedia.org/wiki/AVX-512#CPUs_with_AVX-512); Intel Ice Lake was the first consumer AVX-512 CPU and it is a decent subset of what AMD Zen 4 supports.

mstembera commented:
@87flowers I just noticed that SPLAT_TABLE and SPLAT_PAWN_TABLE are scoped differently.  Is that for a specific reason?
